### PR TITLE
HPCC-12464 Dynamic ESDL U/I

### PR DESCRIPTION
**Pull request by miguelvazq**
Incorporate a GUI for dynamic ESDL.

Signed-off by: Miguel Vazquez miguel.vazquez@lexisnexis.com

### CODE REVIEW

**Comment by miguelvazq** [comment]
@GordonSmith Please review.

**Comment by miguelvazq** [comment]
FYI @JamesDeFabia

**Comment by richardkchapman** [comment]
@GordonSmith

File: esp/src/eclwatch/DynamicESDLDetailsWidget.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        bindingWidgetLoaded: null,
+        definitionWidget: null,
+        definitionWidgetLoaded: false,
+        configurationWidget: null,
+        configurationWidgetLoaded: false,
+        configurationGrid: null,
+
+        postCreate: function (args) {
+            var context = this;
+            this.inherited(arguments);
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLDetailsWidget.js]: Very minor:  always cleaner to have the inherited call as the first line item.

File: esp/src/eclwatch/DynamicESDLDetailsWidget.js [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            });
+            this.dialog = new Dialog({
+                title: context.i18n.PleasePickADefinition,
+                style: "width: 300px;"
+            });
+            this.dialog.addChild(this.definitionDropDown);
+            this.addDefinitionButton = new Button({
+                disabled: false,
+                style: "float:right;",
+                label: "Apply",
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLDetailsWidget.js]: Translation

File: esp/src/eclwatch/DynamicESDLDetailsWidget.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+                                request: {
+                                    EspProcName: context.params.__hpcc_parentName,
+                                    EspBindingName: context.params.Name,
+                                    EsdlDefinitionID: response.GetESDLDefinitionResponse.Id,
+                                    Overwrite: true,
+                                    EsdlServiceName: response.GetESDLDefinitionResponse.ESDLServices.Name,
+                                    Config: "<Methods><Method name='" + response.GetESDLDefinitionResponse.Methods.Method[0].Name + "'/></Methods>"
+                                }
+                            }).then(function (publishresponse) {
+                                if (publishresponse.PublishESDLBindingResponse.status.Code === 0) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLDetailsWidget.js]: Does this need an "exists" test before using it (there might have been an exception)?

File: esp/src/eclwatch/DynamicESDLDetailsWidget.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+                var xml = context.formatXml(context.params.__binding_info.Definition.Interface)
+                context.widget.Definition.init({
+                    sourceMode: "xml"
+                });
+                context.widget.Definition.setText(xml);
+
+                if (!context.widget.Configuration.initalized) {
+                    context.widget.Configuration.init({
+                        Configuration: context.params
+                });
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLDetailsWidget.js]: Minor: indentation

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+                    context.dialog.hide();
+                },
+                label: context.i18n.Save,
+                style: "float: right;"
+            }).placeAt(this.dialog);
+            /*MAY ADD INSTEAD OF AUTOSAVE OPTION
+            this.saveBindingButton = new Button({
+                disabled: true,
+                label: "Save"
+            }).placeAt(this.openButton.domNode, "after");*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Remove commented code.

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+            this.attributeTextBox = new TextBox({
+                label: "Attribute",
+                placeholder: "Attribute",
+                style: "margin-bottom: 10px;"
+
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Blank line

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+                label: context.i18n.Save,
+                style: "float: right;"
+            }).placeAt(this.dialog);
+            /*MAY ADD INSTEAD OF AUTOSAVE OPTION
+            this.saveBindingButton = new Button({
+                disabled: true,
+                label: "Save"
+            }).placeAt(this.openButton.domNode, "after");*/
+            this.attributeTextBox = new TextBox({
+                label: "Attribute",
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Translation?

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+                placeholder: "Attribute",
+                style: "margin-bottom: 10px;"
+
+            }).placeAt(this.dialog);
+            this.valueTextBox = new TextBox({
+                label: "Value",
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Translation?

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+                placeholder: "Value",
+            }).placeAt(this.dialog);
+            this.addAddAttritbuteButton = new Button({
+                disabled: true,
+                onClick: function (val) {
+                    context.dialog.show();
+                },
+                label: "Add Attribute(s)"
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Translation?

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            this.store.mayHaveChildren = function (item) {
+                if (item.Attributes && item.Attributes.Attribute.length) {
+                    return true;
+                }
+                return false;
+            };
+
+            this.store.getChildren = function (parent, options) {
+               var retVal =  this.query({__hpcc_parentName: parent.__hpcc_id}, options);
+               return retVal;
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Minor: no need for retVal;

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+                if (lang.exists("ConfigureESDLBindingMethodResponse.status", response)) {
+                    context.refresh(context.params);
+                }
+            });
+        },
+
+        refreshActionState: function (event) {
+            var context = this;
+            var selection = this.grid.getSelected();
+            var hasSelection = false;
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Not used

File: esp/src/eclwatch/DynamicESDLMethodWidget.js [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+            var isParent = false;
+            var hasParent = false;
+
+            for (var i = 0; i < selection.length; ++i) {
+                hasSelection = true;
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Not used

File: esp/src/eclwatch/DynamicESDLQueryWidget.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+           this.store.mayHaveChildren = function (item) {
+                if (item.__hpcc_parentName) {
+                    return false;
+                }
+                return true;
+            };
+
+            this.store.getChildren = function (parent, options) {
+               var retVal =  this.query({__hpcc_parentName: parent.__hpcc_id}, options);
+               return retVal;
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: Minor: retval not needed

File: esp/src/eclwatch/templates/DynamicESDLDetailsWidget.html [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+                <div id="${id}Toolbar" class="topPanel" data-dojo-props="region: 'top'" data-dojo-type="dijit.Toolbar">
+                    <div id="${id}Refresh" data-dojo-attach-event="onClick:_onRefresh" data-dojo-props="iconClass:'iconRefresh'" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
+                    <span data-dojo-type="dijit.ToolbarSeparator"></span>
+                    <div id="${id}NewPage" class="right" data-dojo-attach-event="onClick:_onNewPage" data-dojo-props="iconClass:'iconNewPage', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.OpenInNewPage}</div>
+                </div>
+                <div id="${id}_Details" data-dojo-props="region: 'center'" data-dojo-type="dijit.layout.ContentPane">
+                    <p>${i18n.PleaseSelectADynamicESDLService}</p>
+                </div>
+            </div>
+            <div id="${id}_Binding" title="Binding" data-dojo-type="dijit.layout.BorderContainer">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/DynamicESDLDetailsWidget.html]: Translation

File: plugins/cassandra/cpp-driver [hunk 15]
```diff
@@ -1 +1 @@
-Subproject commit 468db35af8875930d424395808ddc9523c57c6f2
+Subproject commit d7bbad34db39a51f209c6fadd07c9ec3a0bb86b7
```
Comment (GordonSmith) [inline plugins/cassandra/cpp-driver]: This should not be in your PR

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/DynamicESDLQueryWidget.js (hunk 13 above)
Comment (miguelvazq) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: Tested and children do not show if removed.

File: esp/src/eclwatch/DynamicESDLMethodWidget.js (hunk 10 above)
Comment (miguelvazq) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: Tested and children do not show if removed.

**Comment by richardkchapman** [comment]
@miguelvazq This is yours...

File: esp/src/eclwatch/TargetSelectClass.js [hunk 16]
```diff
@@ -417,6 +420,24 @@ define([
                 }
                 context._postLoad();
             });
-        }
+        },
+
+       loadESDLDefinitions: function () {
+        var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/TargetSelectClass.js]: Indentation

File: esp/src/eclwatch/DynamicESDLQueryWidget.js (hunk 13 above)
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: I meant:  ```return this.query({__hpcc_parentName: parent.__hpcc_id}, options);``` (minor)

File: esp/src/eclwatch/DynamicESDLMethodWidget.js (hunk 10 above)
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLMethodWidget.js]: I meant ```return this.query({__hpcc_parentName: parent.__hpcc_id}, options);``` (minor)

**Review by GordonSmith — APPROVED** [review]
Just minor + indentation comments

**Comment by miguelvazq** [comment]
@GordonSmith updated.

**Comment by miguelvazq** [comment]
Rebased to fix conflicts.

**Review by GordonSmith — APPROVED** [review]
Good to go.

**Comment by richardkchapman** [comment]
Merged manually to resolve conflict

**Pull request closed without merging**